### PR TITLE
#7 Migrate command option parsing from getopt to argparse

### PR DESCRIPTION
**Pull request by BereGabor**
Migrate command option parsing from getopt to argparse
#7 issue

### CODE REVIEW

File: tools/configuration.py [hunk 1]
```diff
@@ -1,24 +1,40 @@
...
-    'authors_top': 5,
-    'commit_begin': '',
-    'commit_end': 'HEAD',
-    'linear_linestats': 1,
-    'project_name': '',
-    'processes': 8,
-    'start_date': '',
-    'output': 'html'
-}
+GITSTAT_VERSION = '0.91'
```
Comment (vifactor) [inline tools/configuration.py]: the version should be fetched from repostat/git_hooks/release_data.json It is automatically updated once I put a git tag of a special format and push to remote.

File: tools/configuration.py [hunk 2]
```diff
@@ -1,24 +1,40 @@
...
+    parser.add_argument('--max_domains', default=10, type=int, help="Display statistic about top active domains")
+    parser.add_argument('--max_ext_length', default=10, type=int, help="Collect statistic from files has shorter "
+                                                                       "extension as max_ext_length")
+    parser.add_argument('--style', default='gitstats.css', type=str, help="css used as generated html result style")
+    parser.add_argument('--max_authors', default=7, type=int, help="Display statistic about top active authors")
+    parser.add_argument('--max_authors_of_months', default=6, type=int, help="Display statistic about top "
+                                                                             "active authors in monthly statistic")
+    parser.add_argument('--authors_top', default=5, type=int,
+                        help="Display statistic about top active authors in total")
+    parser.add_argument('--commit_begin', default="", type=str, help="Analysis starts from this commit date")
```
Comment (vifactor) [inline tools/configuration.py]: commit_(begin/end) are not used and do not think will be: scipt does analysis for whole history

File: tools/configuration.py [hunk 3]
```diff
@@ -1,24 +1,40 @@
...
+
+
+def get_gitstat_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(prog='repo_stat',
+                                     description='Git repository desktop analyzer. Analyze and generate git statistics '
+                                                 'in HTML format, or export into csv files for further analysis.')
+    parser.add_argument('--max_domains', default=10, type=int, help="Display statistic about top active domains")
+    parser.add_argument('--max_ext_length', default=10, type=int, help="Collect statistic from files has shorter "
+                                                                       "extension as max_ext_length")
+    parser.add_argument('--style', default='gitstats.css', type=str, help="css used as generated html result style")
```
Comment (vifactor) [inline tools/configuration.py]: let us keep this hardcoded, do not have plan to make any replacement

File: tools/configuration.py [hunk 4]
```diff
@@ -1,24 +1,40 @@
...
-    'linear_linestats': 1,
-    'project_name': '',
-    'processes': 8,
-    'start_date': '',
-    'output': 'html'
-}
+GITSTAT_VERSION = '0.91'
+
+
+def get_gitstat_parser() -> argparse.ArgumentParser:
```
Comment (vifactor) [inline tools/configuration.py]: You have added all possible options (even unused ones) to the arguments. Let us make it simpler and give user a possibility only to configure important thigs from command line, e.g.

- path to repo
- output path for generated report and data
- version
- maybe path to config file in json-format with all the rest of options

File: tools/configuration.py [hunk 5]
```diff
@@ -1,24 +1,40 @@
...
+                        help="Display name of the project git repo contain. "
+                             "This param currently used in csv output format.")
+    parser.add_argument('--processes', default=8, type=int, help="????")
+    parser.add_argument('--start_date', default='', type=str, help="????")
+    parser.add_argument('--output', default='html', type=str, choices=['html', 'csv'],
+                        help="Statistic output format. Valid values: [html, csv]")
+    parser.add_argument('--version', action='version', version='%(prog)s ' + GITSTAT_VERSION)
+
+    parser.add_argument('git_repo', type=str)
+    parser.add_argument('output_path', default=os.path.abspath(sys.argv[-1]), type=str)
```
Comment (vifactor) [inline tools/configuration.py]: do not provide default here

File: tools/configuration.py [hunk 6]
```diff
@@ -1,24 +1,40 @@
...
+                        help="Display statistic about top active authors in total")
+    parser.add_argument('--commit_begin', default="", type=str, help="Analysis starts from this commit date")
+    parser.add_argument('--commit_end', default="HEAD", type=str, help="Analysis ends on this commit date")
+    parser.add_argument('--linear_linestats', default=1, type=int, help="????")
+    parser.add_argument('--project_name', default="", type=str,
+                        help="Display name of the project git repo contain. "
+                             "This param currently used in csv output format.")
+    parser.add_argument('--processes', default=8, type=int, help="????")
+    parser.add_argument('--start_date', default='', type=str, help="????")
+    parser.add_argument('--output', default='html', type=str, choices=['html', 'csv'],
```
Comment (vifactor) [inline tools/configuration.py]: call it "--format" or "--output_format"

File: tools/configuration.py [hunk 7]
```diff
@@ -71,49 +83,27 @@ def _check_pre_reqs(self):
...
-        outputpath = os.path.abspath(args[-1])
+        args = get_gitstat_parser().parse_args(args_orig)
+
         try:
-            os.makedirs(outputpath)
+            os.makedirs(args.output_path)
         except OSError:
             pass
-        if not os.path.isdir(outputpath):
+        if not os.path.isdir(args.output_path):
```
Comment (vifactor) [inline tools/configuration.py]: check this: https://stackoverflow.com/questions/11415570/directory-path-types-with-argparse for inspiration how path can be validated

File: tools/configuration.py (hunk 4 above)
Comment (BereGabor) [reply tools/configuration.py]: Could I implement this in other PR? Other change request are fixed now.

File: tools/gitstatistics.py [hunk 8]
```diff
@@ -316,7 +316,7 @@ def fetch_recent_activity(self, weeks=None):
             if weeks is None:
                 activity[yyw] = activity.get(yyw, 0) + 1
             elif yyw in weeks:
-                activity[yyw] = activity.get(yyw, 0) + 1
+                actstyleivity[yyw] = activity.get(yyw, 0) + 1
```
Comment (vifactor) [inline tools/gitstatistics.py]: misprint?

Comment (BereGabor) [reply tools/gitstatistics.py]: Wow... soo accurate. Nice :). 
I fixed and thanks

File: tools/configuration.py (hunk 4 above)
Comment (BereGabor) [reply tools/configuration.py]: I add config file feature too in this PR. The solution is come from your recommendation for directory input param validation below :)  It was really useful.

**Pull request merged**